Throw RuntimeException in FluxParser (#421)

dr0i opened the pull request:
If a `RecognitionException` occurs in `FluxParser` it was catched. So callers had
no chance to be aware of the Exception.
As it's not possible to create a new `RecognitionException` with a customized
error message (and throw this new `RecognitionException`) and also to be
backwards compatibel (`RuntimeExceptions` are referred to as unchecked exceptions
and the code would compile just as before) no checked Exception is thrown but
a `RuntimeException` with the customized error message.

Resolves #421.

dr0i commented:
For code reviewers:
The `Flux.g` is used to generate `metafacture-flux/build/generated-src/antlr/main/org/metafacture/flux/parser/FluxParser.java`.

blackwinter reviewed (commented):
I don't understand the specifics of the `catch` block (due to unfamiliarity with ANTLR), so can't really comment on that.

General remarks:

1. Where is the `RecognitionException` caught? I seem to be unable to find it.
2. I'd like to see tests for this error handling.

blackwinter commented:
> Where is the `RecognitionException` caught?

I think I've found it. It's in the generated `FluxParser` source. That's also where you got the error handling that you copied over, right?

dr0i commented:
> unfamiliarity with ANTLR

Yeah, me too :) (traced the stack, found the crux, browsed the net and found the solution. Quite happy debugging :) )

> I think I've found it. It's in the generated FluxParser source. That's also where you got the error handling that you copied over, right?

yep

Re your inline comments:

1.
No - if it's not thrown here it's missing in `flow` where it's thrown to be catchable by calling `FluxCompiler`.

In the `flow` section (the other `catch` block I added) it's not sufficient to just rethrow `RecognitionException` because:

> As it's not possible to create a new RecognitionException with a customized
error message  [...]

The error message of the `RecognitionException` uses character codes (e.g. ASCII `59` instead a `;`) which makes it harder for end users to understand what's wrong. From ANTLRs`RecognitionException`:

> [...] to generally make things as flexible as possible, these exceptions are not created with strings, but rather the information necessary to generate an error [...]

 So_ it's easier for the catcher to produce proper error messages - they let this up to the implementers. Just catching this in `flux` lacks references to produce these messages so it's computed and added to the `RuntimeExcption`.

2.
Tests ... hmhm. Extra work. I don't fear regression here. But you are right, being bold and ignorant is a weak excuse :| ... will add one.

dr0i commented:
@blackwinter added the test

blackwinter commented:
Great :) :+1:

I also understand now why the `RecognitionException` has to be caught and rethrown in `flow`: Otherwise a _default_ `catch` handler would be generated that just swallows the exception. It _would_ be sufficient to _only_ install the `flow` catch handler in order to pass the test, but I've added one to `varDef` as well (and simplified them) so they both end up in the same place. The only thing left to do would be to add the (minimal) catch handler to all the other expressions...

blackwinter commented:
BTW: Would it make sense to throw a `MetafactureException` instead of a `RuntimeException`?

dr0i commented:
Throws now a `FluxParseException` (which extends `MetafactureException`).

>  The only thing left to do would be to add the (minimal) catch handler to all the other expressions...

... 
which (just a note here) could also be done by adding:

```  @rulecatch {
      catch (RecognitionException e) {
          throw e;
      }
  }
```

to the grammar `Flux.g` - but that prevents sometimes more detailed messages (e.g. when doing this with the `atom` expression).
I think this is at least a good start to make many more errors catchable by calling programs (see e.g. https://github.com/metafacture/metafacture-playground/issues/84).

dr0i commented:
Merged this. We can always improve later, if necessary.

blackwinter commented on metafacture-flux/build.gradle:
```diff
@@ -23,6 +23,7 @@ dependencies {
   api project(':metafacture-framework')
   implementation project(':metafacture-commons')
   implementation project(':metafacture-io')
+  implementation project(':metafacture-plumbing')
```
Why was this dependency introduced?

dr0i replied:
This was needed to test the `Tee` command.

blackwinter replied:
I see, but then it should only be a `testRuntime` dependency, right?

blackwinter commented on metafacture-flux/src/main/antlr/org/metafacture/flux/parser/Flux.g:
```diff
@@ -33,16 +33,53 @@ tokens {
+          reportError(e);  // report after consuming so AW sees the token in the exception
+          // we want to return the token we're actually matching
+          Object matchedSymbol = getCurrentInputSymbol(input);
+          input.consume(); // move past ttype token as if all were ok
```
2 x `input.consume()`: Doesn't this consume two tokens?

dr0i replied:
Yeah, good question ... to be honest, I did not totally traced this, just copied the code (as it was before, from the generated class).

blackwinter replied:
Ah, okay.

blackwinter commented on metafacture-flux/src/main/antlr/org/metafacture/flux/parser/Flux.g:
```diff
@@ -33,16 +33,53 @@ tokens {
+          input.consume(); // move past ttype token as if all were ok
+        }
+        // can't recover with single token deletion, try insertion
+        if ( mismatchIsMissingToken(input, follow) ) {
```
Shouldn't this be `else if`?

dr0i replied:
Yeah, I would think so, too.  But this was the originally code from before, and haven't dug into it deeper, maybe something can be a `mismatchIsUnwantedToken` and a `mismatchIsMissingToken` at the same time, don't know. If not, the only impact should be the cost of a fraction of a msecond. I agree in general to avoid unnecessary calls, though.
